src/services/services.ts

```diff
@@ -6259,6 +6379,193 @@ namespace ts {
+                        return referenceEntries && referenceEntries.length ? referenceEntries : undefined;
+                    }
+
+                    // Check if the node is within an extends or implements clause
+                    const containingHeritageClause = getContainingClassHeritageClause(refNode);
+                    if (containingHeritageClause) {
+                        return [getReferenceEntryFromNode(containingHeritageClause.parent)];
+                    }
+
+                    // If we got a type reference, try and see if the reference applies to any expressions that can implement an interface
+                    const containingTypeReference = getContainingTypeReference(refNode);
+                    if (containingTypeReference) {
+                        const parent = containingTypeReference.parent;
+                        if (isVariableLike(parent) && parent.type === containingTypeReference && parent.initializer && isImplementationExpression(parent.initializer)) {
+                            return [getReferenceEntryFromNode(parent.initializer)];
+                        }
+                        else if (isFunctionLike(parent) && parent.type === containingTypeReference && parent.body && parent.body.kind === SyntaxKind.Block) {
+                            let result: ReferenceEntry[];
+
+                            forEachReturnStatement(<Block>parent.body, (returnStatement) => {
+                                if (returnStatement.expression && isImplementationExpression(returnStatement.expression)) {
+                                    (result || (result = [])).push(getReferenceEntryFromNode(returnStatement.expression));
+                                }
+                            });
+
+                            return result;
+                        }
+                        else if (isTypeAssertionExpression(parent) && isImplementationExpression(parent.expression)) {
+                            return [getReferenceEntryFromNode(parent.expression)];
+                        }
+                    }
+                }
+            }
+
+            function getSymbolsForComponentTypes(type: UnionOrIntersectionType, result: Symbol[] = []): Symbol[] {
+                for (const componentType of type.types) {
+                    if (componentType.symbol && componentType.symbol.getFlags() & (SymbolFlags.Class | SymbolFlags.Interface)) {
+                        result.push(componentType.symbol);
+                    }
+                    if (componentType.getFlags() & TypeFlags.UnionOrIntersection) {
+                        getSymbolsForComponentTypes(<UnionOrIntersectionType>componentType, result);
+                    }
+                }
+                return result;
+            }
+
+            function getContainingTypeReference(node: Node): Node {
+                if (node) {

```

Comment:
we should handle all typeNode decedents. e.g.:
        FunctionType,
        ConstructorType,
        TypeQuery,
        ArrayType,
        TupleType,
        UnionType,
        IntersectionType,
        ParenthesizedType,